![](https://www.dii.uchile.cl/wp-content/uploads/2021/06/Magi%CC%81ster-en-Ciencia-de-Datos.png)

# Proyecto: Innovación Tecnológica en Bodoque Bank

**MDS7202: Laboratorio de Programación Científica para Ciencia de Datos**

### Cuerpo Docente:

- Profesor: Pablo Badilla, Ignacio Meza De La Jara
- Auxiliar: Sebastián Tinoco
- Ayudante: Diego Cortez M., Felipe Arias T.

*Por favor, lean detalladamente las instrucciones de la tarea antes de empezar a escribir.*


----

## Reglas

- Fecha de entrega: 01/06/2021
- **Grupos de 2 personas.**
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Estrictamente prohibida la copia. 
- Pueden usar cualquier material del curso que estimen conveniente.

# Motivación

![](https://www.geekmi.news/__export/1621969098810/sites/debate/img/2021/05/25/juan-carlos-bodoque-1.jpg_554688468.jpg)

Juan Carlos Bodoque, el famoso periodista y empresario, decidió diversificar su portafolio de negocios y crear su propio banco. Después de varios años de investigar y analizar el mercado financiero, finalmente logró fundar su entidad bancaria con el objetivo de ofrecer a sus clientes una experiencia personalizada y confiable en sus transacciones financieras. 

Sin embargo, con las nuevas tecnologías, aparecen nuevos desafíos para la joven entidad bancaria. Por ello, Bodoque decide invertir en un equipo de expertos en tecnología y finanzas, para que Bodoque Bank implemente las últimas innovaciones en seguridad y servicio al cliente para garantizar la satisfacción y fidelización de sus clientes.

El primer objetivo de la entidad bancaria será la detección de potenciales clientes fraudulentos, para ello Bodoque Bank les hace entrega de un extenso dataset en el que se registran las actividades que han realizado sus clientes durante los últimos meses. Uno de los puntos que resaltan al pasar el conjunto de datos es que el nombre de los usuarios está protegido y que consideren cada una de las filas como una muestra independiente de la otra.

Tras la solicitud, uno de los mayores accionistas del banco llamado Mario Hugo, les sugiere que al momento de realizar el proyecto tomen las siguientes consideraciones:

- Realicen un análisis exhaustivo de cada variable, ya que los directores tienden a ser quisquillosos con puntos o definiciones sin algún fundamento claro.
- Deben ser muy claros de cómo encontrar los outliers y si es posible expliquen cómo están determinando que un cliente es fraudulento.
- Es muy probable que existan clientes con diferentes comportamientos bancarios, por lo que sería muy buena decisión de separar a los clientes en grupos etarios.


### Definición Formal del Problema

El dataset con el que se trabajará en este primer proyecto será [_The Bank Account Fraud (BAF)_](https://www.kaggle.com/datasets/sgpjesus/bank-account-fraud-dataset-neurips-2022). Este consiste en un dataset para evaluar métodos de detección de fraudes bancarios, el cuál según sus autores es:

- Realista: Se basa en un actualizado dataset de casos reales de detección de fraude bancario.
- Sesgado: Poseen distintos tipos de bias.
- Desbalanceado: Clase positiva extremadamente pequeña.
- Dinámico: Presenta datos temporales y cambios de distribución.
- Preserva la privacidad: Protege a los potenciales clientes a través de la aplicación de privacidad diferencial.

El proyecto tiene por objetivo evaluar los conocimientos adquiridos en la primera mitad del curso, consistente en manejo de datos tabulares (I/O, manipulación, agregaciones, merge y visualizaciones) más la primera parte de modelos consistente en preprocesamiento de datos y detección de anomalías. Por ende, el proyecto consiste en dos tareas principales:

1. Generar un Análisis exploratorio de Datos (EDA) que describa completamente el dataset.
2. Buscar anomalías de forma automatizada.

**Importante:** Esta permitido el uso de librerías externas a las vistas en clases para profundizar aún mas en los análisis. Sin embargo, al momento de utilizar cualqueir metodo deberán explicar que hace y el porque de su aplicación.


---

## Secciones requeridas en la entrega

La siguiente lista detalla las secciones que debe contener su notebook para resolver el proyecto. Es importante que al momento de desarrollar cada una de las secciones, estas sean escritas en un formato tipo **informe**, donde describan detalladamente cada uno de los puntos realizados.

### 1. Introducción [0.5 Puntos]

Elaborar una breve introducción con todo lo necesario para entender qué realizarán durante su proyecto. La idea es que describan de manera formal el proyecto con sus propias palabras y logren describir algunos aspectos básicos tanto del dataset como del análisis a realizar sobre los datos.

Por lo anterior, en esta sección ustedes deberán ser capaces de:

- Describir la tarea asociada al dataset.
- Describir brevemente los datos de entrada que les provee el problema.
- Plantear hipótesis de cómo podrían abordar el problema.

### 2. Lectura y Manejo de Datos [1 Punto]

Este punto tiene por objetivo evaluar la capacidad de leer y manejar distintas fuentes de datos y unirlas.
Para esto, se les provee de 3 datasets distintos: 

- `df_1.parquet` que contiene una cierta cantidad de ejemplos.
- `df_2.parquet` que contiene el resto de los ejemplos.
- `df_email_phone.parquet` que contiene características exclusivas relacionadas a los emails y teléfonos de los ejemplos.

Tanto `df_1` como `df_2` son dataframes que contienen conjuntos (no necesariamente disjuntos) de ejemplos obtenidos a partir de distintas fuentes (la primera más confiable que la segunda)  mientras que `df_email_phone` contiene nuevas características que deben ser fusionadas con el conjunto de ejemplos.
Dicho esto, para este punto se le solicita:

1. Cargar los datasets usando pandas.
2. Explorar superficialmente los datsets por medio de `.head(5)` e `.info()` y reportar número de filas y columnas de cada dataset.
3. Unir los datasets con los conjuntos de ejemplos en un único dataset.
4. Combinar el datasets con ejemplos con las nuevas características usando la estrategia `outer`.
5. Explorar nulos y duplicados de datos (i.e., generar conteos de cada uno y ver algunos de sus ejemplos), explicar su origen y entregar una solución sencilla a estos casos.
6. Limpiar datos y reportar modificaciones a los datasets originales (cuántos ejemplos se eliminaron, cuántos ejemplos quedaron finalmente, etc...).
7. Realizar una segmentación etaria de los datos creando una columna llamada `segmentacion_etaria` para clasificar a los diferentes usuarios. Para la segmentación utilice los siguientes rangos:
   - Joven (Menor a 18 años)
   - Adulto-Joven (18 - 26 años)
   - Adulto (27- 59 años)
   - Persona Mayor (60 años o mas).


### 3. Análisis Exploratorio de Datos [1 Punto]

Esta sección se busca evaluar la capacidad de generar reportes y conclusiones a partir de los patrones observados en estos.

- Utilizando `pandas-profiling` (nombre de la librería: `ydata-profiling`), explore:

    * Analizar cantidad de datos nulos, tipos de datos, duplicados, distribuciones de las variables a través de histogramas.
    * Generar visualizaciones de las interacciones (como por ejemplo, una scatter matrix) en las distintas variables.
    * Ver las correlaciones entre las distintas variables y los valores faltantes de cada una de estas. 
    * Reportar los patrones interesantes observados. Ejemplos:
        - Se observó que las variables numéricas A, B y C siguen una distribución normal / exponencial / logarítmica / no presenta distribución conocida debido a que (usar información sobre el descripción de cada variable)....
        - La mayoría de los ejemplos de las variables I, J y K tienen valor 0.
        - La variable M presenta gran cantidad de outliers.
        - Se observó que las variables categóricas E y F están bien balanceadas / presentan un gran desbalance.
        - Del gráfico de correlaciones se puede inferir que X, Y y Z tienen una correlación alta/baja entre ellas.



### 4. Preparación de Datos [1 Punto]

*Esta sección consiste en generar los distintos pasos para preparar sus datos con el fin de luego poder crear su modelo.*

Dentro de los aspectos minimos que se les solicitarán en este desarrollo estan:

- Definir preprocesadores para datos categóricos y ordinales.
- Setear las transformaciones en un `ColumnTransformer`.
- Transformar todo el datset.

**Notas**: 

- Las variables `id` y `fraud_bool` no deben ser transformadas, pero si pasadas a la siguiente etapa. (Estudiar la transformación `passthrough`).
- La salida de la transformación debe ser un pandas dataframe. Para esto, investiguen acerca de la nueva [`set_output` API](https://scikit-learn.org/stable/auto_examples/miscellaneous/plot_set_output.html). Si se levanta un error por matriz sparse, desactiven su uso en la transformación correspondiente.

### 5. Visualización en baja dimensionalidad [1 Punto]

Programar un script para proyectar los datos en baja dimensionalidad usando `PCA` o `UMAP`.

**Notas**: 

- Para realizar las proyecciones no consideren la variable `fraud_bool`, ya que esto provocaría un _data leakage_ (i.e., usar información del futuro para analizar el mismo fenómeno que quieren descubrir).
- Utilice un muestreo significativo del dataframe (10k a 25k ejemplos) para generar las proyecciones. De lo contrario, probablemente el proyector no sea capaz de asignar todos los recursos necesarios y lance alguna excepción.

Luego, por cada segmento etario generado en la sección 2, implemente un gráfico de dispersión (idealmente en plotly express) la proyección en 2D generada y coloree cada punto según la etiqueta `fraud_bool`.

Reportar si existen patrones y/o relaciones interesantes con respecto a variables de interés.


### 6. Explorar Anomalías [1.5 Puntos]

Proponer una técnica para detectar clientes fraudulentos en base a detección de anomalías. 
Al igual que el punto anterior, la técnica que proponga debe ser aplicada por separado en los diferentes segmentos etarios descritos en el punto 2 (es decir, debe entrenar un detector de anomalías para cada uno de estos grupos). 

Luego, al igual que el punto anterior, genere un gráfico de dispersión con las proyecciones 2D en donde el coloreado sea ahora las etiquetas predichas por el detector de anomalías.

Por último, usando queries de numpy/pandas, calcule dos ratios (P y R):

- P: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de datos predichos como fraude.
- R: Cantidad de ejemplos predichos correctamente como fraude / cantidad total de ejemplos que eran realmente fraude.


De los resultados obtenidos responda las siguientes preguntas:

- ¿Qué significan los ratios y sus valores?
- ¿Qué tan correctas fueron las predicciones realizadas por su modelo según los ratios calculados?
- ¿Son coherentes los resultados obtenidos?.
- ¿Cómo se comportan los casos de fraudes para los diferentes rangos etarios?

Justifique cada una de sus respuestas.

---

## Esquema de la Tarea


Pueden usar el siguiente esquema para organizar la tarea (y borrar todo lo anterior).
Obviamente **no deben limitarse a lo que está escrito en esta**: puede incrementar en caso de más técnicas y obviar algunas partes en caso que alguna y otro punto no aplique a su problema.

Pueden borrar las instrucciones anteriores y quedarse solo con lo que viene a continuación.

---
<br>

<br>

<br>



# Proyecto

### Equipo:

- \<Primer integrante\>
- \<Segundo integrante\>


### Link de repositorio de GitHub: `\<http://....\>`




## 1. Introducción

La predicción de fraude en las empresas en un problema conocido y bastante estudiado debido al interés que supone sobre minizar los riesgos que adoptan las empresas con sus clientes. El problema que se presenta en el proyecto consta de detección de fraude de cliente por parte de banco. 

La forma de abordar el problema consistirá de emplear modelos de clasificación (donde 1 correspondería a cliente clasificado como fraudulento y 0 como cliente no fraudulento) en los diferentes segmentos etarios de la empresa. Los cuales son 4:

* Joven (Menor a 18 años)
* Adulto-Joven (18 - 26 años)
* Adulto (27- 59 años)
* Persona Mayor (60 años o mas).

Donde los datos de entrada del modelo corresponden a datos asociados al cliente donde algunas columnas tienen datos nulos y sus representaciones corresponden a valores negativos, -1 o valores NA. 

En cuanto a la hipótesis del problema, se usará un enfoque donde el aprendizaje de máquinas tiene todo el poder predictivo, o sea en otras palabras, se le entregarán todas las variables que cuenta el dataset y el modelo de ML será el que le de el peso a las variables que son más importantes para predecir este problema. En cuanto a lso resultados esperados, no se espera que sean muy ambiciosos debido a que es un típico problema donde hay mucho desbalance de clases y los modelos tenderán a predecir la clase mayoritaria.

## 2. Lectura y Manejo de Datos




### 2.1 Cargar datos en dataframes

In [1]:
#Conectamos el colab a drive para cargar los archivos
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [2]:
#Importamos las bases de datos
import pandas as pd

df1=pd.read_parquet('/content/drive/MyDrive/lab_progra/df_1.parquet', engine='pyarrow')
df2=pd.read_parquet('/content/drive/MyDrive/lab_progra/df_2.parquet', engine='pyarrow')
df_email_phone=pd.read_parquet('/content/drive/MyDrive/lab_progra/df_email_phone.parquet', engine='pyarrow')

### 2.2 Exploración inicial

In [3]:
print('Las dimensiones de df_1 son: '+str(df1.shape[0])+' filas por '+str(df1.shape[1])+' columnas')
print('')
print('Las dimensiones de df_2 son: '+str(df2.shape[0])+' filas por '+str(df2.shape[1])+' columnas')
print('')
print('Las dimensiones de df_email_phone son: '+str(df_email_phone.shape[0])+' filas por '+str(df_email_phone.shape[1])+' columnas')

Las dimensiones de df_1 son: 389782 filas por 29 columnas

Las dimensiones de df_2 son: 747410 filas por 29 columnas

Las dimensiones de df_email_phone son: 934730 filas por 7 columnas


Dado el código anterior, se sabe que el primer dataframe df_1 tiene 389782 filas y 29 columnas, el dataframe df_2 tiene 747410 filas y 29 columnas y por último, el dataframe df_email_phone tiene 934730 filas y 7 columnas.

In [4]:
df1.head(5)

id  fraud_bool  income  prev_address_months_count  \
104446  729517           0     0.7                         -1   
269483  149585           0     0.8                         -1   
4102     64486           0     0.9                         -1   
351767  825283           0     0.5                         -1   
126377    8308           0     0.9                         -1   

        current_address_months_count  customer_age  days_since_request  \
104446                           305            60            0.030059   
269483                           140            50            0.015659   
4102                             171            50            0.001409   
351767                            85            30            0.027292   
126377                            39            30            0.010945   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
104446               -1.599455           AC           990  ...   
269483                3.951994           AA          1269  ...   
4102                 28.159779           AB          4430  ...   
351767               -1.310498           AB          1698  ...   
126377               -1.450972           AC           569  ...   

        proposed_credit_limit  foreign_request    source  \
104446                  500.0                0  INTERNET   
269483                  200.0                0  INTERNET   
4102                   1500.0                0  INTERNET   
351767                  200.0                0  INTERNET   
126377                  200.0                0  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
104446                   8.865992   windows                   0   
269483                   4.654872     linux                   1   
4102                     3.720953     linux                   1   
351767                   2.912670   windows                   1   
126377                   2.283680     other                   1   

       device_fraud_count  month        x1        x2  
104446                  0      2 -0.245425  0.568811  
269483                  0      7  0.009336 -2.096682  
4102                    0      1  2.229616 -0.005823  
351767                  0      3 -0.193945  0.861207  
126377                  0      3 -2.440650  0.354986  

[5 rows x 29 columns]

In [5]:
df2.head(5)

id  fraud_bool  income  prev_address_months_count  \
699954  303612           0     0.8                         -1   
701206  568066           0     0.9                         -1   
270304  893344           0     0.7                         -1   
346538  247688           0     0.8                         -1   
147297  674478           0     0.8                         -1   

        current_address_months_count  customer_age  days_since_request  \
699954                            21            40            0.010837   
701206                            70            50            0.711485   
270304                           133            50            0.006856   
346538                           207            50            0.026194   
147297                            24            50            0.012646   

        intended_balcon_amount payment_type  zip_count_4w  ...  \
699954               -0.853976           AD           239  ...   
701206               25.137456           AA          1873  ...   
270304               -0.815330           AC           675  ...   
346538               -0.736615           AB           999  ...   
147297               10.744253           AA          3198  ...   

        proposed_credit_limit  foreign_request    source  \
699954                  200.0                0  INTERNET   
701206                  200.0                0  INTERNET   
270304                 1000.0                0  INTERNET   
346538                  200.0                0  INTERNET   
147297                  500.0                1  INTERNET   

        session_length_in_minutes device_os  keep_alive_session  \
699954                  48.127390     other                   0   
701206                   8.792273   windows                   1   
270304                  15.916225     other                   1   
346538                   0.509942     linux                   1   
147297                  10.562945     other                   0   

       device_fraud_count  month        x1        x2  
699954                  0      3  0.236931  0.077061  
701206                  0      3  0.571536  1.151563  
270304                  0      7  0.224252 -0.590459  
346538                  0      6  0.648528 -0.315911  
147297                  0      6  0.041475  0.383332  

[5 rows x 29 columns]

En cuanto a las características que tienen los dataframe similares df_1 y df_2 se puede ver que tienen diferentes variables asociadas al cliente y también si es que el cliente hizo fraude o no. Estas se enumeran a continuación en conjunto a su significado:

* income (numeric):  Ingreso anual del solicitante (en forma decil). Rangos entre [0.1, 0.9].


* prev_address_months_count (numeric): Número de meses en el domicilio registrado anterior del solicitante, es decir, la residencia anterior del solicitante, si corresponde. Varía entre [−1, 380] meses (-1 es un valor faltante).

* current_address_months_count (numeric): meses en la dirección registrada actual del solicitante. Varía entre [−1, 429] meses (-1 es un valor faltante).

* customer_age (numeric): Edad del solicitante en años, redondeada a la década. Rangos entre [10, 90] años.

* days_since_request (numeric): N número de días transcurridos desde que se realizó la solicitud. Rangos entre [0, 79] días.

* intended_balcon_amount (numeric):  Monto inicial transferido para la aplicación. Varía entre [−16, 114] (los negativos son valores faltantes).

* payment_type (categorical): tipo de plan de pago de crédito. 5 valores posibles (anonimizados).

* zip_count_4w (numeric): Number of applications within same zip code in last 4 weeks.
Ranges between [1, 6830].

* velocity_6h (numeric): Velocidad de aplicaciones totales realizadas en las últimas 6 horas, es decir, número promedio de aplicaciones por hora en las últimas 6 horas. Varía entre [−175, 16818].

* velocity_24h (numeric): Velocidad de aplicaciones totales realizadas en las últimas 24 horas, es decir, número promedio de aplicaciones por hora en las últimas 24 horas. Rangos entre [1297, 9586]

* velocity_4w (numeric):  Velocidad de aplicaciones totales realizadas en las últimas 4 semanas, es decir, promedio de aplicaciones por hora en las últimas 4 semanas. Varía entre [2825, 7020].

* bank_branch_count_8w (numeric):  Número de solicitudes totales en la sucursal bancaria seleccionada en las últimas 8 semanas. Varía entre [0, 2404].


* employment_status (categorical):  Situación laboral del solicitante. 7 valores posibles (anonimizados).

* credit_risk_score (numeric): Puntuación interna del riesgo de la aplicación. Varía entre [−191, 389].


* housing_status (categorical): Estado residencial actual del solicitante. 7 valores posibles (anonimizados).


* bank_months_count (numeric): ¿Cuántos años tiene la cuenta anterior (si se tiene) en meses? Varía entre [−1, 32] meses (-1 es un valor faltante).

* has_other_cards (binary): Si el solicitante tiene otras tarjetas de la misma compañía bancaria.

* proposed_credit_limit (numeric): límite de crédito propuesto por el solicitante. Rangos entre [200, 2000].

* foreign_request (binary): si el país de origen de la solicitud es diferente del país del banco.

* source (categorical): Fuente en línea de aplicación. Ya sea navegador (INTERNET) o app (TELEAPP).

* session_length_in_minutes (numeric):  duración de la sesión del usuario en el sitio web bancario en minutos. Varía entre [−1, 107] minutos (-1 es un valor faltante).

* device_os (categorical): Sistema operativo del dispositivo que realizó la solicitud. Los valores posibles son: Windows, macOS, Linux, X11 u otros.

* keep_alive_session (binary): opción de usuario al cerrar la sesión.


* device_fraud_count (numeric):  Número de aplicaciones fraudulentas con dispositivo usado. Rangos entre [0, 1].

* month (numeric): Mes en que se realizó la solicitud. Rangos entre [0, 7].

* fraud_bool (binary): Si la aplicación es fraudulenta o no.

In [6]:
df_email_phone.head(5)

id  name_email_similarity  date_of_birth_distinct_emails_4w  \
987231  624588               0.555653                                15   
79954   620810               0.849718                                 2   
567130  580633               0.110898                                 3   
500891    6379               0.671250                                15   
55399   366511               0.772932                                 7   

        email_is_free  device_distinct_emails_8w  phone_home_valid  \
987231              0                          1                 0   
79954               0                          1                 0   
567130              0                          1                 1   
500891              0                          1                 0   
55399               0                          1                 0   

        phone_mobile_valid  
987231                   1  
79954                    0  
567130                   0  
500891                   1  
55399                    1

Y en cuanto a la base de datos df_email_phone, las columnas que tiene son:
* name_email_similarity (numeric): Métrica de similitud entre el correo electrónico y el nombre del solicitante. Los valores más altos representan una mayor similitud. Rangos entre [0, 1].
* date_of_birth_distinct_emails_4w (numeric): Número de correos electrónicos para solicitantes con la misma fecha de nacimiento en las últimas 4 semanas. Rangos entre [0, 39].
* email_is_free (binary): Dominio del correo electrónico de la aplicación (ya sea gratuito o de pago).
* phone_home_valid (binary): Validez del teléfono de casa proporcionado.
* phone_mobile_valid (binary): Validez del teléfono móvil proporcionado.
* device_distinct_emails (numeric): Número de correos electrónicos distintos en el sitio web bancario desde el dispositivo usado en las últimas 8 semanas. Varía entre [−1, 2] correos electrónicos (-1 es un valor faltante).

In [7]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 389782 entries, 104446 to 193209
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            389782 non-null  int64  
 1   fraud_bool                    389782 non-null  int64  
 2   income                        389782 non-null  float64
 3   prev_address_months_count     389782 non-null  int64  
 4   current_address_months_count  389782 non-null  int64  
 5   customer_age                  389782 non-null  int64  
 6   days_since_request            389782 non-null  float64
 7   intended_balcon_amount        389782 non-null  float64
 8   payment_type                  389782 non-null  object 
 9   zip_count_4w                  389782 non-null  int64  
 10  velocity_6h                   389782 non-null  float64
 11  velocity_24h                  389782 non-null  float64
 12  velocity_4w                   389782 no

In [8]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 747410 entries, 699954 to 384611
Data columns (total 29 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   id                            747410 non-null  int64  
 1   fraud_bool                    747410 non-null  int64  
 2   income                        747410 non-null  float64
 3   prev_address_months_count     747410 non-null  int64  
 4   current_address_months_count  747410 non-null  int64  
 5   customer_age                  747410 non-null  int64  
 6   days_since_request            747410 non-null  float64
 7   intended_balcon_amount        747410 non-null  float64
 8   payment_type                  747410 non-null  object 
 9   zip_count_4w                  747410 non-null  int64  
 10  velocity_6h                   747410 non-null  float64
 11  velocity_24h                  747410 non-null  float64
 12  velocity_4w                   747410 no

In [9]:
df_email_phone.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 934730 entries, 987231 to 986114
Data columns (total 7 columns):
 #   Column                            Non-Null Count   Dtype  
---  ------                            --------------   -----  
 0   id                                934730 non-null  int64  
 1   name_email_similarity             934730 non-null  float64
 2   date_of_birth_distinct_emails_4w  934730 non-null  int64  
 3   email_is_free                     934730 non-null  int64  
 4   device_distinct_emails_8w         934730 non-null  int64  
 5   phone_home_valid                  934730 non-null  int64  
 6   phone_mobile_valid                934730 non-null  int64  
dtypes: float64(1), int64(6)
memory usage: 57.1 MB


### 2.3 Unir dataframes

In [10]:
#Se agrega una columna para tener la trazabilidad de los orígenes de los datos
df1['origen']='base 1'
df2['origen']='base 2'

In [11]:
df_concat=pd.concat([df1,df2]).reset_index(drop=True)

### 2.4 Unir nuevas variables

In [12]:
bbdd_final = pd.merge(df_concat, df_email_phone, on='id', how='left')

In [13]:
bbdd_final.shape

(1137192, 36)

In [14]:
bbdd_final.groupby('device_os').size()

device_os
linux        383185
macintosh     57016
other        341352
windows      346379
x11            9260
dtype: int64

Al unir las variables se verifica cómo se ven los valores de las columnas con el objetivo de ver algún valor extraño. Primero se verán las columans numericas y luego las categoricas,

In [15]:
bbdd_final.describe().transpose()

count           mean            std  \
id                                1137192.0  500113.268421  288722.419048   
fraud_bool                        1137192.0       0.011010       0.104347   
income                            1137192.0       0.578548       0.288240   
prev_address_months_count         1137192.0      14.647000      42.981190   
current_address_months_count      1137192.0      99.175138      94.063764   
customer_age                      1137192.0      41.343828      13.766449   
days_since_request                1137192.0       0.901719       5.000559   
intended_balcon_amount            1137192.0       8.560989      20.532602   
zip_count_4w                      1137192.0    1518.532112     966.147032   
velocity_6h                       1137192.0    5491.808343    2940.041218   
velocity_24h                      1137192.0    4661.668206    1451.255984   
velocity_4w                       1137192.0    4734.725941     871.270566   
bank_branch_count_8w              1137192.0     201.271150     473.813362   
credit_risk_score                 1137192.0     139.263714      71.452421   
bank_months_count                 1137192.0      11.148524      12.128297   
has_other_cards                   1137192.0       0.249646       0.432809   
proposed_credit_limit             1137192.0     551.614890     506.588620   
foreign_request                   1137192.0       0.023880       0.152675   
session_length_in_minutes         1137192.0       7.817620       8.242504   
keep_alive_session                1137192.0       0.556244       0.496827   
device_fraud_count                1137192.0       0.000000       0.000000   
month                             1137192.0       3.656215       2.116896   
x1                                1137192.0       0.014586       1.013109   
x2                                1137192.0       0.011012       1.012876   
name_email_similarity             1057465.0       0.487641       0.291457   
date_of_birth_distinct_emails_4w  1057465.0       7.778368       4.814321   
email_is_free                     1057465.0       0.518114       0.499672   
device_distinct_emails_8w         1057465.0       1.022282       0.191573   
phone_home_valid                  1057465.0       0.493414       0.499957   
phone_mobile_valid                1057465.0       0.856897       0.350178   

                                           min            25%            50%  \
id                                0.000000e+00  249918.500000  500347.500000   
fraud_bool                        0.000000e+00       0.000000       0.000000   
income                            1.000000e-01       0.300000       0.600000   
prev_address_months_count        -1.000000e+00      -1.000000      -1.000000   
current_address_months_count     -1.000000e+00      27.000000      64.000000   
customer_age                      1.000000e+01      30.000000      50.000000   
days_since_request                3.112791e-08       0.007434       0.015659   
intended_balcon_amount           -1.553733e+01      -1.179949      -0.834103   
zip_count_4w                      1.000000e+00     886.000000    1209.000000   
velocity_6h                      -1.741097e+02    3337.138253    5192.389354   
velocity_24h                      1.322325e+03    3504.029907    4641.080855   
velocity_4w                       2.870592e+03    4238.872451    4813.757967   
bank_branch_count_8w              0.000000e+00       1.000000      10.000000   
credit_risk_score                -1.770000e+02      90.000000     130.000000   
bank_months_count                -1.000000e+00       1.000000       6.000000   
has_other_cards                   0.000000e+00       0.000000       0.000000   
proposed_credit_limit             1.900000e+02     200.000000     200.000000   
foreign_request                   0.000000e+00       0.000000       0.000000   
session_length_in_minutes        -1.000000e+00       3.152311       5.248896   
keep_alive_session                0.000000e+00    

de las variables numericas se puede observar que la variable device_fraud_count no tiene variabilidad por lo que se debe eliminar

In [16]:
bbdd_final.describe(include = [object]).transpose()

count unique       top     freq
payment_type       1137192      5        AB   453427
employment_status  1137192      7        CA   778603
housing_status     1137192      7        BC   381659
source             1137192      2  INTERNET  1128411
device_os          1137192      5     linux   383185
origen             1137192      2    base 2   747410

In [17]:
print(bbdd_final['payment_type'].unique())
print(bbdd_final['employment_status'].unique())
print(bbdd_final['source'].unique())
print(bbdd_final['device_os'].unique())
print(bbdd_final['housing_status'].unique())

['AC' 'AA' 'AB' 'AD' 'AE']
['CC' 'CF' 'CA' 'CB' 'CE' 'CD' 'CG']
['INTERNET' 'TELEAPP']
['windows' 'linux' 'other' 'macintosh' 'x11']
['BB' 'BA' 'BE' 'BC' 'BD' 'BF' 'BG']


Y en cuanto a las variables categoricas se observa que hay valores normales en su data.

### 2.5 Verificar nulos y duplicados, generar explicaciones sobre sus fuentes y proponer soluciones.



#### Nulos

Primero se transformarán los valores nulos respectivos por cada columna donde la forma de representar los nulos correspondía a -1 o a valores negativos, luego se cuenta la cantidad de nulos que hay por columna. Se puede observar que las columnas que se agregaron por el cruce son las que tienen estos valores nulos, además de las columnas, prev_address_months_count, intended_balcon_amount, current_address_months_count, bank_months_count, session_length_in_minutes, payment_tyoe, employment_status, source, device_os y housing_status.

In [18]:
import numpy as np
#transformar a valores nulos

bbdd_final['prev_address_months_count']=bbdd_final['prev_address_months_count'].replace(-1, np.NaN) 
bbdd_final['current_address_months_count']=bbdd_final['current_address_months_count'].replace(-1, np.NaN) 
bbdd_final[bbdd_final['intended_balcon_amount'] < 0] = 0
bbdd_final['intended_balcon_amount']=bbdd_final['intended_balcon_amount'].replace(0, np.NaN) 
bbdd_final['bank_months_count']=bbdd_final['bank_months_count'].replace(-1, np.NaN) 
bbdd_final['session_length_in_minutes']=bbdd_final['session_length_in_minutes'].replace(-1, np.NaN) 
bbdd_final['device_distinct_emails_8w']=bbdd_final['device_distinct_emails_8w'].replace(-1, np.NaN)   


In [19]:
bbdd_final.isnull().sum()

id                                       0
fraud_bool                               0
income                                   0
prev_address_months_count           219072
current_address_months_count           458
customer_age                             0
days_since_request                       0
intended_balcon_amount              854466
payment_type                             0
zip_count_4w                             0
velocity_6h                              0
velocity_24h                             0
velocity_4w                              0
bank_branch_count_8w                     0
employment_status                        0
credit_risk_score                        0
housing_status                           0
bank_months_count                     4800
has_other_cards                          0
proposed_credit_limit                    0
foreign_request                          0
source                                   0
session_length_in_minutes              170
device_os  

Lo que se realizará por cada columna con nulos es lo siguiente:


1.   Para variables que tienen muchas columnas con nulos, estas se eliminan: caso de prev_address_months_count y de intended_balcon_amount, payment_tyoe, employment_status, source, device_os y housing_status.
2.   Para las últimas columnas agregadas por el cruce, se investigará el origen de los nulos.
3. Para el resto de las  columnas, se eligirá la media como valor para imputar esos valores.








**De aquí en adelante se investigará el origen de las últimas columnas:**
Por la cantidad repetida de nulos en cada columna, una hipótesis es que todos esos valores nulos provienen de los mismos registros. Esto se comprueba a continuación, donde primero se toma una columna cualquiera, luego se filtra por las filas nulas y se vuelve a calcular la cantidad de nulos que hay por cada columna. Si da la misma cantidad, esto significa que los nulos provienen de los mismos registros.

In [20]:
nulos=bbdd_final[bbdd_final['name_email_similarity'].isnull()]

In [21]:
nulos.isnull().sum()

id                                      0
fraud_bool                              0
income                                  0
prev_address_months_count           15218
current_address_months_count           31
customer_age                            0
days_since_request                      0
intended_balcon_amount                  0
payment_type                            0
zip_count_4w                            0
velocity_6h                             0
velocity_24h                            0
velocity_4w                             0
bank_branch_count_8w                    0
employment_status                       0
credit_risk_score                       0
housing_status                          0
bank_months_count                     306
has_other_cards                         0
proposed_credit_limit                   0
foreign_request                         0
source                                  0
session_length_in_minutes              14
device_os                         

Como dio la misma cantidad de nulos que en la base donde no se filtro, quiere decir que todos los nulos provienen de los mismos registros.

Para saber la razón de los nulos, se separá la base en una que contiene sólo los registros con nulos y la otra en una base que sólo contenga datos sin nulos, luego de esta separación se verificarán 2 cosas. Primero, ver si hay diferencias notables en los estadísticos de ambas bases. Esto con el objetivo de ver si hay alguna diferencia en las poblaciones de ambas bases que pueda llevar a pensar que hay una variable que se correlaciona con la nulidad de algunos datos.

Lo segundo que se verificará es si el origen de los datos puede llegar a afectar la nulidad, debido a que df1 y df2 tienen 2 orígenes distintos y además, los datos de df2 no son tan confiables, según enunciado, pero no sabemos el por qué.

##### Estadísticos

Ya habíamos separado una base solo con nulos, así que falta armar la base sin nulos. Luego de crearla, se verán las diferencias entre los estadísticos de ambas bases.

In [22]:
sin_nulos=bbdd_final[bbdd_final['name_email_similarity'].notnull()]

In [23]:
sin_nulos.describe().transpose()

count           mean            std  \
id                                1117507.0  117686.697215  254143.895555   
fraud_bool                        1117507.0       0.001200       0.034620   
income                            1117507.0       0.141831       0.289232   
prev_address_months_count          913653.0       4.363502      23.936543   
current_address_months_count      1117080.0      24.699417      64.212955   
customer_age                      1117507.0       9.591904      18.572273   
days_since_request                1117507.0       0.298132       2.450397   
intended_balcon_amount             263041.0      37.487422      24.101325   
zip_count_4w                      1117507.0     360.739588     793.096679   
velocity_6h                       1117507.0    1373.263456    2862.674949   
velocity_24h                      1117507.0    1142.096176    2175.899396   
velocity_4w                       1117507.0    1131.910085    2083.678364   
bank_branch_count_8w              1117507.0      72.605174     304.599536   
credit_risk_score                 1117507.0      30.907177      65.558820   
bank_months_count                 1113013.0       3.745851       8.604605   
has_other_cards                   1117507.0       0.083798       0.277085   
proposed_credit_limit             1117507.0     127.596561     330.855171   
foreign_request                   1117507.0       0.005895       0.076554   
session_length_in_minutes         1117351.0       1.929197       5.453145   
keep_alive_session                1117507.0       0.125255       0.331008   
device_fraud_count                1117507.0       0.000000       0.000000   
month                             1117507.0       0.828241       1.814717   
x1                                1117507.0       0.002205       0.488171   
x2                                1117507.0       0.000626       0.487167   
name_email_similarity             1117507.0       0.117578       0.256214   
date_of_birth_distinct_emails_4w  1117507.0       1.825606       4.011428   
email_is_free                     1117507.0       0.117529       0.322049   
device_distinct_emails_8w         1117471.0       0.238456       0.435621   
phone_home_valid                  1117507.0       0.116810       0.321194   
phone_mobile_valid                1117507.0       0.210670       0.407784   

                                         min        25%        50%        75%  \
id                                  0.000000   0.000000   0.000000   0.000000   
fraud_bool                          0.000000   0.000000   0.000000   0.000000   
income                              0.000000   0.000000   0.000000   0.000000   
prev_address_months_count           0.000000   0.000000   0.000000   0.000000   
current_address_months_count        0.000000   0.000000   0.000000   0.000000   
customer_age                        0.000000   0.000000   0.000000   0.000000   
days_since_request                  0.000000   0.000000   0.000000   0.000000   
intended_balcon_amount              0.000027  20.767272  32.973873  49.734067   
zip_count_4w                        0.000000   0.000000   0.000000   0.000000   
velocity_6h                        -5.946966   0.000000   0.000000   0.000000   
velocity_24h                        0.000000   0.000000   0.000000   0.000000   
velocity_4w                         0.000000   0.000000   0.000000   0.000000   
bank_branch_count_8w                0.000000   0.000000   0.000000   0.000000   
credit_risk_score                -157.000000   0.000000   0.000000   0.000000   
bank_months_count                   0.000000   0.000000   0.000000   0.000000   
has_other_cards                     0.000000   0.000000   0.000000   0.000000   
proposed_credit_limit               0.000000   0.000000   0.000000   0.000000   
foreign_request                     0.000000   0.000000   0.000000   0.000000   
session_length_in_minutes           0.000000   0.000000   0.000000   0.000000   
keep_alive_session            

In [24]:
nulos.describe().transpose()

count           mean            std  \
id                                19685.0  497914.650800  290185.947358   
fraud_bool                        19685.0       0.004572       0.067464   
income                            19685.0       0.595682       0.279218   
prev_address_months_count          4467.0      66.739199      68.053631   
current_address_months_count      19654.0     105.151216      95.606371   
customer_age                      19685.0      40.803658      14.046411   
days_since_request                19685.0       1.229434       4.948818   
intended_balcon_amount            19685.0      37.649522      24.042736   
zip_count_4w                      19685.0    1539.557633     946.065046   
velocity_6h                       19685.0    5818.891787    2987.248922   
velocity_24h                      19685.0    4872.745097    1467.858247   
velocity_4w                       19685.0    4810.487521     879.149690   
bank_branch_count_8w              19685.0     315.155601     572.390005   
credit_risk_score                 19685.0     130.898298      71.216811   
bank_months_count                 19379.0      16.205480      10.936813   
has_other_cards                   19685.0       0.356566       0.478997   
proposed_credit_limit             19685.0     542.041656     485.210348   
foreign_request                   19685.0       0.025959       0.159017   
session_length_in_minutes         19671.0       8.268987       8.862318   
keep_alive_session                19685.0       0.535484       0.498752   
device_fraud_count                19685.0       0.000000       0.000000   
month                             19685.0       3.512624       2.126321   
x1                                19685.0       0.011218       1.005061   
x2                                19685.0       0.004628       1.010925   
name_email_similarity                 0.0            NaN            NaN   
date_of_birth_distinct_emails_4w      0.0            NaN            NaN   
email_is_free                         0.0            NaN            NaN   
device_distinct_emails_8w             0.0            NaN            NaN   
phone_home_valid                      0.0            NaN            NaN   
phone_mobile_valid                    0.0            NaN            NaN   

                                           min            25%            50%  \
id                                3.700000e+01  242860.000000  499223.000000   
fraud_bool                        0.000000e+00       0.000000       0.000000   
income                            1.000000e-01       0.400000       0.700000   
prev_address_months_count         8.000000e+00      27.000000      34.000000   
current_address_months_count      0.000000e+00      33.000000      68.000000   
customer_age                      1.000000e+01      30.000000      40.000000   
days_since_request                3.839882e-07       0.007429       0.016044   
intended_balcon_amount            7.248630e-04      21.080925      32.996352   
zip_count_4w                      4.000000e+00     922.000000    1259.000000   
velocity_6h                       5.005369e+01    3557.217670    5435.900465   
velocity_24h                      1.530362e+03    3744.905810    4880.607464   
velocity_4w                       2.992104e+03    4262.966872    4874.283139   
bank_branch_count_8w              0.000000e+00       4.000000      15.000000   
credit_risk_score                -1.390000e+02      81.000000     124.000000   
bank_months_count                 1.000000e+00       5.000000      16.000000   
has_other_cards                   0.000000e+00       0.000000       0.000000   
proposed_credit_limit             2.000000e+02     200.000000     200.000000   
foreign_request                   0.000000e+00       0.000000       0.000000   
session_length_in_minutes         2.088441e-02       3.482721       5.494952   
keep_alive_session                0.000000e+00       0.000000       1.000000   
device_fraud_count            

In [25]:
resta1=sin_nulos.describe().transpose()

In [26]:
resta2=nulos.describe().transpose()

In [27]:
resta1-resta2

count           mean           std  \
id                                1097822.0 -380227.953585 -36042.051803   
fraud_bool                        1097822.0      -0.003372     -0.032843   
income                            1097822.0      -0.453851      0.010015   
prev_address_months_count          909186.0     -62.375696    -44.117089   
current_address_months_count      1097426.0     -80.451799    -31.393416   
customer_age                      1097822.0     -31.211753      4.525862   
days_since_request                1097822.0      -0.931302     -2.498421   
intended_balcon_amount             243356.0      -0.162100      0.058589   
zip_count_4w                      1097822.0   -1178.818045   -152.968367   
velocity_6h                       1097822.0   -4445.628330   -124.573973   
velocity_24h                      1097822.0   -3730.648921    708.041149   
velocity_4w                       1097822.0   -3678.577436   1204.528674   
bank_branch_count_8w              1097822.0    -242.550427   -267.790469   
credit_risk_score                 1097822.0     -99.991121     -5.657990   
bank_months_count                 1093634.0     -12.459629     -2.332208   
has_other_cards                   1097822.0      -0.272768     -0.201912   
proposed_credit_limit             1097822.0    -414.445095   -154.355177   
foreign_request                   1097822.0      -0.020064     -0.082463   
session_length_in_minutes         1097680.0      -6.339790     -3.409174   
keep_alive_session                1097822.0      -0.410229     -0.167744   
device_fraud_count                1097822.0       0.000000      0.000000   
month                             1097822.0      -2.684383     -0.311604   
x1                                1097822.0      -0.009013     -0.516891   
x2                                1097822.0      -0.004002     -0.523757   
name_email_similarity             1117507.0            NaN           NaN   
date_of_birth_distinct_emails_4w  1117507.0            NaN           NaN   
email_is_free                     1117507.0            NaN           NaN   
device_distinct_emails_8w         1117471.0            NaN           NaN   
phone_home_valid                  1117507.0            NaN           NaN   
phone_mobile_valid                1117507.0            NaN           NaN   

                                           min            25%            50%  \
id                               -3.700000e+01 -242860.000000 -499223.000000   
fraud_bool                        0.000000e+00       0.000000       0.000000   
income                           -1.000000e-01      -0.400000      -0.700000   
prev_address_months_count        -8.000000e+00     -27.000000     -34.000000   
current_address_months_count      0.000000e+00     -33.000000     -68.000000   
customer_age                     -1.000000e+01     -30.000000     -40.000000   
days_since_request               -3.839882e-07      -0.007429      -0.016044   
intended_balcon_amount           -6.974734e-04      -0.313653      -0.022479   
zip_count_4w                     -4.000000e+00    -922.000000   -1259.000000   
velocity_6h                      -5.600065e+01   -3557.217670   -5435.900465   
velocity_24h                     -1.530362e+03   -3744.905810   -4880.607464   
velocity_4w                      -2.992104e+03   -4262.966872   -4874.283139   
bank_branch_count_8w              0.000000e+00      -4.000000     -15.000000   
credit_risk_score                -1.800000e+01     -81.000000    -124.000000   
bank_months_count                -1.000000e+00      -5.000000     -16.000000   
has_other_cards                   0.000000e+00       0.000000       0.000000   
proposed_credit_limit            -2.000000e+02    -200.000000    -200.000000   
foreign_request                   0.000000e+00       0.000000       0.000000   
session_length_in_minutes        -2.088441e-02      -3.482721      -5.494952   
keep_alive_session                0.000000e+00       0.000000      -1.000000   
d

Al ver las diferencias que pueden haber entre los estadísticos de ambas bases, se puede ver que las mayores diferencias se dan en las columnas: zip_count_4w, velocity_6h, velocity_24h, velocity_4w y bank_branch_count_8w. Al observar los valores de esas columnas en los estadísticos de las bases respectivas, se puede observar que la magnitud más alta en comparación a las otras columnas se debe al rango en donde se mueven esas variables, siendo estos mucho más amplios que el resto de las variables. 
Por lo que se concluye que no hay diferencias significativas entre ambos estadisticos. Por lo que se explorará si el origen de los nulos proviene de los orígenes de las bases.

##### Origen

In [28]:
nulos.groupby(['origen'])['origen'].count()

origen
base 2    19685
Name: origen, dtype: int64

In [29]:
sin_nulos.groupby(['origen'])['origen'].count()

origen
0         854466
base 1     96928
base 2    166113
Name: origen, dtype: int64

Se puede observar que los datos nulos provienen solo del origen de la base 2. Esto se puede deber a la confiabilidad de los datos que se informa en el inicio del proyecto, donde mencionan la confiabilidad de ambas bases: **"la primera más confiable que la segunda"**. No mencionan más sobre la confiabilidad de los datos, pero con la información que tenemos se puede suponer que la confiabilidad de los datos de la segunda base puede tener relación a que en ella no se pudieron recolectar datos de contacto del cliente por un motivo aleatorio, esto debido a que no se observó que alguna variable extra del dataset tuviera relación con la nulidad de los datos.

Como este tipo de nulo es un missing at random, la solución más óptima es dropear los registros que contengan los nulos.

#### Duplicados


Primero se verá cuántos datos son los duplicados en la base y luego, mostrarán los casos que se repiten en la base de datos

In [30]:
bbdd_final=bbdd_final.drop(['origen'], axis=1)
len(bbdd_final[bbdd_final.duplicated(keep=False)])

929306

In [31]:
pd.set_option('display.max_columns', None)

In [32]:
bbdd_final[bbdd_final.duplicated(keep=False)]

id  fraud_bool  income  prev_address_months_count  \
0             0           0     0.0                        0.0   
1        149585           0     0.8                        NaN   
3             0           0     0.0                        0.0   
4             0           0     0.0                        0.0   
5        891636           0     0.6                        NaN   
...         ...         ...     ...                        ...   
1137186       0           0     0.0                        0.0   
1137187   28360           0     0.8                        NaN   
1137188       0           0     0.0                        0.0   
1137190       0           0     0.0                        0.0   
1137191       0           0     0.0                        0.0   

         current_address_months_count  customer_age  days_since_request  \
0                                 0.0             0            0.000000   
1                               140.0            50            0.015659   
3                                 0.0             0            0.000000   
4                                 0.0             0            0.000000   
5                               137.0            60            0.046158   
...                               ...           ...                 ...   
1137186                           0.0             0            0.000000   
1137187                         137.0            30            0.018466   
1137188                           0.0             0            0.000000   
1137190                           0.0             0            0.000000   
1137191                           0.0             0            0.000000   

         intended_balcon_amount payment_type  zip_count_4w  velocity_6h  \
0                           NaN            0             0     0.000000   
1                      3.951994           AA          1269  2844.260015   
3                           NaN            0             0     0.000000   
4                           NaN            0             0     0.000000   
5                     25.841925           AA          1512  7669.494145   
...                         ...          ...           ...          ...   
1137186                     NaN            0             0     0.000000   
1137187               49.088586           AA           915  4814.292418   
1137188                     NaN            0             0     0.000000   
1137190                     NaN            0             0     0.000000   
1137191                     NaN            0             0     0.000000   

         velocity_24h  velocity_4w  bank_branch_count_8w employment_status  \
0            0.000000     0.000000                     0                 0   
1         2789.827355  3141.137362                  1605                CF   
3            0.000000     0.000000                     0                 0   
4            0.000000     0.000000                     0                 0   
5         4251.257682  5721.121146                   271                CC   
...               ...          ...                   ...               ...   
1137186      0.000000     0.000000                     0                 0   
1137187   3219.381648  5444.343899                    19                CA   
1137188      0.000000     0.000000                     0                 0   
1137190      0.000000     0.000000                     0                 0   
1137191      0.000000     0.000000                     0                 0   

         credit_risk_score housing_status  bank_months_count  has_other_cards  \
0                        0              0                0.0                0   
1                      105             BA                9.0                1   
3                        0              0                0.0                0   
4                        0              0                0.0                0   
5                       33             BB               28.0                

Como cada registro representa la situación de un cliente específico, que hayan registros con la misma información es incorrecto y podría inducir a sesgos en el análisis, por lo que la forma de solucionar este problema es dropeando los registros duplicados.

### 2.6 Limpiar

Se eliminan de la base los registros duplicados 
device_fraud_count

In [33]:
bbdd_final=bbdd_final.drop_duplicates()

In [34]:
bbdd_final.shape

(232158, 35)

Se eliminan las columnas que tienen muchos NA y la variable device_fraud_count que no presenta variacion

In [35]:
bbdd_final=bbdd_final.drop(['prev_address_months_count','intended_balcon_amount','device_fraud_count'], axis=1)

Se eliminan las filas que tienen las últimas columnas con NA y también el id 0 debido a que parece ser un dato de prueba que no significa nada

In [36]:
bbdd_final=bbdd_final[bbdd_final['name_email_similarity'].notnull()]
bbdd_final = bbdd_final[bbdd_final['id'] != 0]

Además de eliminar esos datos, se eliminan las columnas x1 y x2, ya que no significan nada dentro de la base de datos.

In [37]:
bbdd_final=bbdd_final.drop(['x1','x2'], axis=1)

Por último, las otras varibles que presentaban valores menores de nulos se les imputaran esos valores por la mdedia del resto en caso de ser numerica y en caso de ser categorica, la moda


In [38]:
bbdd_final['current_address_months_count'] = bbdd_final['current_address_months_count'].fillna(bbdd_final['current_address_months_count'].mean())
bbdd_final['bank_months_count'] = bbdd_final['bank_months_count'].fillna(bbdd_final['bank_months_count'].mean())
bbdd_final['session_length_in_minutes'] = bbdd_final['session_length_in_minutes'].fillna(bbdd_final['session_length_in_minutes'].mean())
bbdd_final['device_distinct_emails_8w'] = bbdd_final['device_distinct_emails_8w'].fillna(bbdd_final['device_distinct_emails_8w'].mean().round())                             

In [39]:
bbdd_final.isnull().sum()

id                                  0
fraud_bool                          0
income                              0
current_address_months_count        0
customer_age                        0
days_since_request                  0
payment_type                        0
zip_count_4w                        0
velocity_6h                         0
velocity_24h                        0
velocity_4w                         0
bank_branch_count_8w                0
employment_status                   0
credit_risk_score                   0
housing_status                      0
bank_months_count                   0
has_other_cards                     0
proposed_credit_limit               0
foreign_request                     0
source                              0
session_length_in_minutes           0
device_os                           0
keep_alive_session                  0
month                               0
name_email_similarity               0
date_of_birth_distinct_emails_4w    0
email_is_fre

### 2.7 Nueva columna con segmentaciones

In [40]:
import numpy as np
bins = [0, 18, 26, 59, np.inf]
names = ['Menor a 18', '19 - 26 años', '27 - 59 años', '60 años o mas']

bbdd_final['segmentacion_etaria'] = pd.cut(bbdd_final['customer_age'], bins, labels=names)

In [41]:
bbdd_final.shape

(217104, 31)

## 3. Análisis Exploratorio de Datos

In [42]:
!pip install ydata_profiling

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [43]:
#from ydata_profiling import ProfileReport

#profile = ProfileReport(bbdd_final, title="EDA")
#profile.to_file("eda.html")

### 3.1 Análisis del EDA

* has_another_card: la variable posee 2 posibles valores, 0 y 1, donde cuando toma el valor 0 corresponde a que no tiene otras tarjetas y cuando es 1 significa que sí posee otras tarjetas. Los valores están desbalanceados, donde el valor 1 corresponde a menos de 1/3 de los valores que corresponden a 0.

 En cuanto a las correlaciones con otras variables, no se observa que tenga una correlacion significativa con otras variables, se puede observar que ronda una correlacion cercana a 0. Por último, no hay valores nulos enla variable.

* proposer_credit_limit: la variable posee 12 valores distintos, a pesar de estar definida como una variable flotante. En cuanto a su distribución, no se observa una distribución clara de los datos. y casi el 58% de los datos corresponde a la variable 200. Por otro lado, los valores altos de esta variable no parecen ser outliers debido a que presentan cierta frecuencia, esta no es con valores menores a 10. Po último, no se observan valores nulos en estas variables.

 En cuanto a losestadísticos de esta variable, la media ronda por 551, y las medidas de dispersión dan cuenta de una alta dispersión de los datos. Esto tiene sentido mirando el histograma debido a que no se observa que se concentren los datos en algún lugar.

 En cuanto a las correlaciones con otras variables, credit_risk_score presenta el mayor valor de correlacion con esta variable teniendo un orden de 0.6. En cuanto a las otras variables, no pasan de una correlacion de 0.2.

* foreign_request: esta variable tiene solo 2 tipos de valores, 0 y 1, donde el valor 1 quiere decir que la solicitud proviene de un pais distinta al banco y cuando es 0, el pais es el mismo del banco. Las clases en esta variable están muy desbalanceadas, donde el valor 0 corresponde a 97,6% de los datos de esta variable. En cuanto a las correlaciones de esta variable, esta no presenta variables correlacionadas distintas a valores cercanos a 0. Tampoco se presentan valores nulos en la variable.

* source: esta variable tiene solo 2 valores, internet y teleapp, donde la variable internet corresponde al 99% de los valores de esa variable. En cuanto a las correlaciones de esa variable, la variable device_distinct_emails_8w	tiene una correlación igual a 0.419 freente a las otras variables donde casi son 0. En cuanto a los valores nulos, no se encuentran valores nulos en esta variable.

* session_length_in_minutes: es una variable numérica, que presenta una distribucion log-normal, con valores más acumulados cercano a 0. Posee una alta varianza en sus datos y no presenta valores nulos . En cuanto a los valores outliers, se puede apreciar que sí es una variable que los presenta debido a que tiene valores que están en la cola superior de los datos. En cuando a las correlaciones entre los datos, se puede observar que las que presentan mayor correlación no superan el valor de 0.2.

* device_os: esta variable tiene 5 valores posibles, other, linux, windows, macintosh y x11. Donde la calse mayoritaria, que es other presenta un 37% de los datos. Esta variable no tiene valores nulos y no presenta correlaciones entre otras variables.

* keep_alive_session: es una variable que presenta 2 valores posibles, 1 y 0, donde la clase mayoritaria es 1 que corresponde al 53% de los datos. Esta variable no presenta valores nulos y tampoco presenta correlaciones con otroas variables.

* month: es una variable que no presenta una distribución clara, es algo oscilante donde en el mes 3 y 6 hay picos. En cuanto a las correlaciones que presenta con las otras variables, es hasta el momento la que presenta más correlaciones entre las variables. Esta variable presenta correlaciones negativas con las variables: zip_count_4w, velocity_6h, velocity_24h y  velocity_4w, donde la última presenta un valor de -0.8 de la otra variable. Tampoco presenta nulos la variable.

* name_email_similarity: la distribución de la variable no tiene una distribución clara, tiene cierto caracter sinusoidal donde se dispara en el valor de 1, donde se acumula la moda de la variable. En cuanto a las correlaciones que tiene esa variable con otras variables, se puede observar que no presenta correlaciones muy altas con otras variables, de hecho la más alta no supera el módulo de 0.2. Esta variable no presenta valores nulos.

* date_of_birth_distinct_emails_4w: esta variable parece tener una distribución bimodal con los 2 picos presentes en 5 y en 12. Presenta mucha variacion entre sus datos y en cuanto a la correlacion con otras variables, la variable customer age presenta una correlacion igual a -0.5 y otras correlaciones menor alrededores de 0.2 en modulo. ESta variable no presenta valores nulos.

* email_is_free: esta variable presenta 2 valores, 0 y 1 y su distribuciones son parecidas, alrededor de 50% cada una. En cuanto a las correlaciones con otras variables, no superan el modulo de 0.2. Esta variable no presenta valores nulos.

* device_distinct_emails_8w: los posibles valores de esta variable son 0, 1 y 2, siendo la clase mayoritaria la clase 1, siendo el 97% de los casos. En cuanto a las correlaciones entre las otras variables, no presenta correlaciones altas. Tampoco presenta valores nulos la variable.

* phone_home_valid: los posibles valores de esta variable son 2, que son 1 y 0. ambos tienen alrededor de 50% de los datos. En cuanto a las correlaciones que pueden existir con las otras variables, no presenta altas correlaciones, siendo la mayoria bajo al modulo de 0.2.

* phone_mobile_valid: los posibles valores de esta variable son 1 y 0. donde la clase mayoritaria es 1 que está presente en el 89% de los datos. En cuando a las correlaciones entre las otras variables, no presenta correlaciones fuertes con las otras, de hecho presenta valores menores del modelo de 0.2.

* segmentacion_etaria: presenta 4 valores posibles donde la clase mayoritaria presenta el 71% de los datos. En cuanto a la correlacion entre otras variables, tiene sentido q esté altamente correlacionado con el customer_age ya que se construyó a partir de esa variable esa variable. No presenta variables nulas.








## 4. Preprocesamiento

### 4.1 Declarar `ColumnTransformer`

# ELIMINAR VARIABLES X1 X2 Y LAS QUE ELIMINÉ EN LA SECCIÓN DE LIMPIEZA, REVISALA.

In [44]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.compose import make_column_transformer
from sklearn.preprocessing import OneHotEncoder, OrdinalEncoder, StandardScaler, MinMaxScaler, PowerTransformer

In [45]:
import sklearn

In [46]:
bbdd_final.columns

Index(['id', 'fraud_bool', 'income', 'current_address_months_count',
       'customer_age', 'days_since_request', 'payment_type', 'zip_count_4w',
       'velocity_6h', 'velocity_24h', 'velocity_4w', 'bank_branch_count_8w',
       'employment_status', 'credit_risk_score', 'housing_status',
       'bank_months_count', 'has_other_cards', 'proposed_credit_limit',
       'foreign_request', 'source', 'session_length_in_minutes', 'device_os',
       'keep_alive_session', 'month', 'name_email_similarity',
       'date_of_birth_distinct_emails_4w', 'email_is_free',
       'device_distinct_emails_8w', 'phone_home_valid', 'phone_mobile_valid',
       'segmentacion_etaria'],
      dtype='object')

In [47]:
variables_numericas=[
    'income',
    'prev_address_months_count',
    'current_address_months_count',
    'customer_age',
    'days_since_request',
    'intended_balcon_amount',
    'zip_count_4w',
    'velocity_6h',
    'velocity_24h',
    'velocity_4w',
    'bank_branch_count_8w',
    'credit_risk_score',
    'bank_months_count',
    'proposed_credit_limit',
    'session_length_in_minutes',
    'month',
    'name_email_similarity',
    'date_of_birth_distinct_emails_4w'
]
variables_ordinales=[
    'payment_type',
    'employment_status',
    'housing_status',
    'has_other_cards',
    'foreign_request',
    'source',
    'device_os',
    'keep_alive_session',
    'device_fraud_count',
    'origen',
    'email_is_free',
    'device_distinct_emails_8w',
    'phone_home_valid',
    'phone_mobile_valid',
    'segmentacion_etaria'
] 

In [48]:
numerical_transformer = StandardScaler()
ordinal_transformer = OneHotEncoder(sparse_output=False)

In [49]:
#Realizamos One Hot Encoder a las variablees ordinales
column_transformer = ColumnTransformer(
    transformers=[ 
      ('ord_transform',ordinal_transformer, variables_ordinales),
      ('num_transform',numerical_transformer, variables_numericas),
      ('no_transform', 'passthrough', ['id', 'fraud_bool'])     
    ],
    remainder='passthrough'
)

### 4.2 Transformar datos

In [50]:
transformed_data = column_transformer.fit_transform(bbdd_final)

ValueError: ignored

In [ ]:
new_columns = column_transformer.named_transformers_['ord_transform'].get_feature_names_out(variables_ordinales)
new_columns_2 = column_transformer.named_transformers_['num_transform'].get_feature_names_out(variables_numericas)
all_columns = list(new_columns) + list(new_columns_2) + ['id', 'fraud_bool']

df_encoded = pd.DataFrame(transformed_data, columns=all_columns)

In [ ]:
df_encoded.describe()

In [ ]:
df_encoded.columns.values

In [ ]:
bbdd_final['fraud_bool'].describe()

In [ ]:
df_encoded[df_encoded['fraud_bool']>0].describe

## 5. Visualización en Baja Dimensionalidad

### 5.1 Muestrear dataframe

In [ ]:
from sklearn.decomposition import PCA

In [ ]:
muestreo = df_encoded.sample(n=10000, replace=True, random_state=190423)
caract = df_encoded.columns.drop('fraud_bool', 'id')

pca = PCA(n_components=2)
transformed_data = pca.fit_transform(muestreo[caract])

In [ ]:
transformed_data.shape

In [ ]:
muestreo['comp1'] = transformed_data[:,0]
muestreo['comp2'] = transformed_data[:,1]

In [ ]:
clases=['segmentacion_etaria_19 - 26 años',
       'segmentacion_etaria_27 - 59 años', 'segmentacion_etaria_60 años o mas',
       'segmentacion_etaria_Menor a 18']

### 5.2 Proyectar y agregar proyecciones al dataframe de muestreo

**ACÁ HAY UN PROBLEMA, SI SE HACE LA SUMA DE LAS 4 BASES, NO SUMAN 10000**


In [ ]:
menor = muestreo[muestreo['segmentacion_etaria_Menor a 18']>0][['comp1','comp2', 'fraud_bool']]
joven = muestreo[muestreo['segmentacion_etaria_19 - 26 años']>0][['comp1','comp2', 'fraud_bool']]
adulto = muestreo[muestreo['segmentacion_etaria_27 - 59 años']>0][['comp1','comp2', 'fraud_bool']]
mayor = muestreo[muestreo['segmentacion_etaria_60 años o mas']>0][['comp1','comp2', 'fraud_bool']]

### 5.3 Visualizar según rangos etarios

In [ ]:
import plotly.express as px

fig = px.scatter(menor, x='comp1', y='comp2', color='fraud_bool')
fig.show()

In [ ]:
fig = px.scatter(joven, x='comp1', y='comp2', color='fraud_bool')
fig.show()

In [ ]:
fig = px.scatter(adulto, x='comp1', y='comp2', color='fraud_bool')
fig.show()

In [ ]:
fig = px.scatter(mayor, x='comp1', y='comp2', color='fraud_bool')
fig.show()

## 6. Anomalías

### 6.1 Implementar detector de anomalías sobre dataframe de muestreo

# ACÁ SELECCIONA LOS MODELOS SEGÚN EL RECALL/PRECISION MEJOR. NOMBRO LOS 2 PORQUE SE ME CONFUNDEN, UNO DE ELLOS FAVORECE A LA CLASE POSITIVA QUE ES LO QUE QUEREMOS VER. ME PARECE Q ES EL RECALL EL Q SIRVE

La forma en que se detectaran anomalías por rango etario serán usando modelos de clasificación. Se probarán base dummy,decisiontree y knn. Los modelos que presenten un mejor F1-Score serán los elegidos. 

In [ ]:
from sklearn.dummy import DummyClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier 
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix

In [ ]:
from sklearn.metrics import f1_score, recall_score, precision_score
from sklearn.model_selection import train_test_split
import numpy as np

def run_classifier(clf, X, y):
        metrics = {'f1-score': [], 'precision': [], 'recall': []}
    
        #dividiendo los datos de entrenamiento y validación
        X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=.30,  random_state=101)

        clf.fit(X_train,y_train) #fiteamos el modelo
        predictions = clf.predict(X_test) #Generamos las predicciones
       
        
        ### FIN COMPLETAR ACÁ
        
        metrics['y_pred'] = predictions
        metrics['f1-score'].append(f1_score(y_test, predictions)) 
        metrics['recall'].append(recall_score(y_test, predictions))
        metrics['precision'].append(precision_score(y_test, predictions))
        matriz= confusion_matrix(y_test, predictions)


        return metrics, matriz

#### Menor

In [ ]:
X=menor[['comp1','comp2']]
y=menor['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified', random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c2 = ("KNN", KNeighborsClassifier(n_neighbors=10))

classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

Acá se elige el modelo de base dummy

In [ ]:
X=menor[['comp1','comp2']]
y=menor['fraud_bool']

X_train_menor, X_test_menor, y_train_menor, y_test_menor = train_test_split(X, y, test_size=.30,  random_state=101)
clf_menor=DummyClassifier(strategy='stratified', random_state=10)
clf_menor.fit(X_train_menor,y_train_menor) #fiteamos el modelo
predictions_menor = clf_menor.predict(X_test_menor) #Generamos las predicciones
predictions_menor_s=pd.DataFrame(predictions_menor).reset_index()
menor_2 =  pd.concat([X_test_menor.reset_index(), predictions_menor_s], axis=1)
confusion_matrix(y_test_menor, predictions_menor)

#### Joven

In [ ]:
# Ejecutar este código

X=joven[['comp1','comp2']]
y=joven['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified', random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c2 = ("KNN", KNeighborsClassifier(n_neighbors=10))

classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

Acá se elige el modelo de base dummy

In [ ]:
X=joven[['comp1','comp2']]
y=joven['fraud_bool']

X_train_joven, X_test_joven, y_train_joven, y_test_joven = train_test_split(X, y, test_size=.30,  random_state=101)
clf_joven=DummyClassifier(strategy='stratified', random_state=10)
clf_joven.fit(X_train_joven,y_train_joven) #fiteamos el modelo
predictions_joven = clf_joven.predict(X_test_joven) #Generamos las predicciones
predictions_joven_s=pd.DataFrame(predictions_joven).reset_index()
joven_2 =  pd.concat([X_test_joven.reset_index(), predictions_joven_s], axis=1)
confusion_matrix(y_test_joven, predictions_joven)

#### Adulto

In [ ]:
# Ejecutar este código

X=adulto[['comp1','comp2']]
y=adulto['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified',  random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c3 = ("KNN", KNeighborsClassifier(n_neighbors=10))


classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

Acá se elige el modelo de base dummy

In [ ]:
X=adulto[['comp1','comp2']]
y=adulto['fraud_bool']

X_train_adulto, X_test_adulto, y_train_adulto, y_test_adulto = train_test_split(X, y, test_size=.30,  random_state=101)
clf_adulto=DummyClassifier(strategy='stratified', random_state=10)
clf_adulto.fit(X_train_adulto,y_train_adulto) #fiteamos el modelo
predictions_adulto = clf_adulto.predict(X_test_adulto) #Generamos las predicciones
predictions_adulto_s=pd.DataFrame(predictions_adulto).reset_index()
adulto_2 =  pd.concat([X_test_adulto.reset_index(), predictions_adulto_s], axis=1)
confusion_matrix(y_test_adulto, predictions_adulto)

#### Mayor

In [ ]:
# Ejecutar este código

X=mayor[['comp1','comp2']]
y=mayor['fraud_bool']

c0 = ("Base Dummy", DummyClassifier(strategy='stratified',  random_state=10))
c1 = ("Decision Tree", DecisionTreeClassifier(max_depth=5))
c2 = ("KNN", KNeighborsClassifier(n_neighbors=10))


classifiers = [c0, c1, c2]

results = {}
for name, clf in classifiers:
    metrics, matriz = run_classifier(clf, X, y)   # hay que implementarla en el bloque anterior.
    results[name] = metrics
    print("----------------")
    print("Resultados para clasificador: ", name) 
    print(matriz)
    print("Precision promedio:", np.array(metrics['precision']).mean())
    print("Recall promedio:", np.array(metrics['recall']).mean())
    print("F1-score promedio:", np.array(metrics['f1-score']).mean())
    print("----------------\n\n")  

Acá se elige el modelo de base dummy en cada segmentación debido a que es el modelo que entrega los mejores F1-Score.

In [ ]:
X=mayor[['comp1','comp2']]
y=mayor['fraud_bool']

X_train_mayor, X_test_mayor, y_train_mayor, y_test_mayor = train_test_split(X, y, test_size=.30,  random_state=101)
clf_mayor=DummyClassifier(strategy='stratified', random_state=10)
clf_mayor.fit(X_train_mayor,y_train_mayor) #fiteamos el modelo
predictions_mayor = clf_mayor.predict(X_test_mayor) #Generamos las predicciones
predictions_mayor_s=pd.DataFrame(predictions_mayor).reset_index()
mayor_2 =  pd.concat([X_test_mayor.reset_index(), predictions_mayor_s], axis=1)
confusion_matrix(y_test_mayor, predictions_mayor)

### 6.2 Agregar resultados a dataframe de muestreo




En la base de datos que se construirá, se dejarán los valores de la matriz de confusión de cada uno de los segmentos con el modelo dummy.

In [ ]:
# initialize list of lists
data = [['menor',279, 239,208,198], ['joven', 341, 244, 238, 182], ['adulto', 29,41,35,46],['mayor',325,219,210,148]]
  
# Create the pandas DataFrame
df = pd.DataFrame(data, columns=['segmento', 'TP', 'FP','FN','TN'])
  
# print dataframe.
df

### 6.3 Visualizar según rangos etarios

#### Menor

In [ ]:
fig = px.scatter(menor_2, x='comp1', y='comp2', color=0)
fig.show()

####Joven

In [ ]:
fig = px.scatter(joven_2, x='comp1', y='comp2', color=0)
fig.show()

####Adulto

In [ ]:
fig = px.scatter(adulto_2, x='comp1', y='comp2', color=0)
fig.show()

####Mayor

In [ ]:
fig = px.scatter(mayor_2, x='comp1', y='comp2', color=0)
fig.show()

### 6.4 Calcular ratios y responder

In [ ]:
...